tests/cpp-runtime/opencl/opencl_texture_pool_test.cc

```diff
@@ -26,126 +26,271 @@
-  EXPECT_EQ(item.first, 1024);
-  EXPECT_EQ(item.second, 768);
-
-  pool.Alloc(t->device, workspace, 64, 12455, type);
-  EXPECT_EQ(pool.AllocatedListSize(), 2);
-  EXPECT_EQ(pool.FreeListSize(), 0);
-  item = pool.AllocatedListItemSize(1);
-  EXPECT_EQ(item.first, 64);
-  EXPECT_EQ(item.second, 12455);
-
-  pool.Free(data1);
-  EXPECT_EQ(pool.AllocatedListSize(), 1);
-  EXPECT_EQ(pool.FreeListSize(), 1);
-  item = pool.AllocatedListItemSize(0);
-  EXPECT_EQ(item.first, 64);
-  EXPECT_EQ(item.second, 12455);
-  item = pool.FreeListItemSize(0);
-  EXPECT_EQ(item.first, 1024);
-  EXPECT_EQ(item.second, 768);
-
-  pool.Alloc(t->device, workspace, 768, 1024, type);
-  EXPECT_EQ(pool.AllocatedListSize(), 2);
-  EXPECT_EQ(pool.FreeListSize(), 0);
-  item = pool.AllocatedListItemSize(0);
-  EXPECT_EQ(item.first, 64);
-  EXPECT_EQ(item.second, 12455);
-  item = pool.AllocatedListItemSize(1);
-  EXPECT_EQ(item.first, 1024);
-  EXPECT_EQ(item.second, 1024);
+
+  // Allocate two buffers and check list sizes at each stage
+  auto did = t->device.device_id;
+  void* buf_data1 = pool.AllocMemory(t->device, 1024, type);
+  EXPECT_EQ(pool.AllocatedListSize(did), 1);
+  EXPECT_EQ(pool.FreeListSize(did), 0);
+  void* buf_data2 = pool.AllocMemory(t->device, 2048, type);
+  EXPECT_EQ(pool.AllocatedListSize(did), 2);
+  EXPECT_EQ(pool.FreeListSize(did), 0);
+  FreeDataSpace(workspace, buf_data1, t->device, pool);
+  EXPECT_EQ(pool.AllocatedListSize(did), 1);
+  EXPECT_EQ(pool.FreeListSize(did), 1);
+  FreeDataSpace(workspace, buf_data2, t->device, pool);
+  EXPECT_EQ(pool.AllocatedListSize(did), 0);
+  EXPECT_EQ(pool.FreeListSize(did), 2);
 }
 
-TEST(OpenCLTexturePool, avoid_reusing_too_big_textures) {
+TEST(OpenCLTexturePool, buffer_allocations_and_reuse) {
   OpenCLWorkspace* workspace = OpenCLWorkspace::Global();
   OpenCLThreadEntry* t = workspace->GetThreadEntry();
-  PoolWrapper pool;
-  EXPECT_EQ(pool.AllocatedListSize(), 0);
-  EXPECT_EQ(pool.FreeListSize(), 0);
+  MemoryPoolWrapper pool(t->device, workspace);
+  DLDataType type{kDLFloat, 16, 1};
+
+  // Request buffer of 1024
+  auto did = t->device.device_id;
+  void* buf_data1 = pool.AllocMemory(t->device, 1024, type);
+  EXPECT_EQ(pool.AllocatedListSize(did), 1);
+  EXPECT_EQ(pool.FreeListSize(did), 0);
+  // Request buffer of 2048
+  void* buf_data2 = pool.AllocMemory(t->device, 2048, type);
+  EXPECT_EQ(pool.AllocatedListSize(did), 2);
+  EXPECT_EQ(pool.FreeListSize(did), 0);
+  // Free buffer 1 (1024)
+  FreeDataSpace(workspace, buf_data1, t->device, pool);
+  EXPECT_EQ(pool.AllocatedListSize(did), 1);
+  EXPECT_EQ(pool.FreeListSize(did), 1);
 
+  // Request buffer of 768
+  void* buf_data3 = pool.AllocMemory(t->device, 768, type);
+  EXPECT_EQ(pool.AllocatedListSize(did), 2);  // Should have reused the 1024 slot
+  EXPECT_EQ(pool.FreeListSize(did), 0);
+
+  // Free buffer 2 (2048)
+  FreeDataSpace(workspace, buf_data2, t->device, pool);
+  EXPECT_EQ(pool.AllocatedListSize(did), 1);
```

Comment:
I would suggest checking the size of the allocated buffers.

Reply:
Make sense, as there are reallocation possible in some cases.